### PR TITLE
Add corrected unicode support

### PR DESCRIPTION
**Pull request by Artoria2e5**
This PR provides complete support for Unicode. As mentioned in #7's comment area, wcwidth dependency is added.

### CODE REVIEW

**Comment by Arthur2e5** [comment]
Shoot. PEP 8 is not happy with docstrings like:

```Python
def foo():
    """
    blah.
    """
```

Only
```Python
def foo():
        """
        blah.
        """
```
makes it happy, which is almost nonsense.

**Comment by Arthur2e5** [comment]
Now travis CI is mad. https://github.com/Arthur2e5/Pylsy/blob/pr-7/pylsy/pylsy.py#L60 does use `self.`

**Comment by Artoria2e5** [comment]
By the way I believe I have found out what the _alignment is hard_ saying means. With the current padding scheme (which uses a looks-weird-to-me `maxwidth * 2` colwidth), one would expect such output:

```
  一二三四五   |
   不打老虎    |
```

Here the two-col wide Chinese chars are not aligned. Still it doesn't look that bad…

Anyway, this is **not** a reason not to include correct Unicode & wchar support.

File: AUTHORS [hunk 1]
```diff
@@ -5,4 +5,5 @@ Leviathan <Leviathan1995@outlook.com>
 OldPanda <zjh0930@gmail.com>
 Sinux <wshmelo@gmail.com>
 Umair Hussain <suhussai@ualberta.ca>
-孙海清 <Leviathan@sunhaiqingdeMacBook-Air.local>
```
Comment (Artoria2e5) [inline AUTHORS:8]: Looks like a git mistake, so fixed..

**Comment by leviathan0992** [comment]
呃。。还是无法对齐阿，我当初就是因为无法对齐才close的，我自己也试过添加中文

File: pylsy/pylsy.py [hunk 2]
```diff
@@ -81,31 +76,38 @@ def _print_head(self):
     def _print_value(self):
         for line in range(self.Lines_num):
             for col, length in zip(self.Table, self.AttributesLength):
+                vals = list(col.values())[0]
+                val = vals[line] if len(vals) != 0 and line < len(vals) else ''
                 self.StrTable += "| "
-                value_length = length * 2 - 1
```
Comment (Arthur2e5) [inline pylsy/pylsy.py]: I don't know why you wrote so much crap here…

**Comment by Artoria2e5** [comment]
居中的情况排版还需要对齐？

```
|  那我要  |
|   逼死   |
| 强迫症了 |
```

**Comment by leviathan0992** [comment]
我那行代码的意思就是为了居中，然后计算value开始的位置。

**Comment by leviathan0992** [comment]
我是居中，然后对齐。中文看起来和英文一样美观，但是后来发现太麻烦了，而且平台不一样，中文的宽度还不同

**Comment by Artoria2e5** [comment]
……我是说为了居中需要那么多 crap 还到处复制粘贴吗……read my diff.

```
+ - - + - - - + - - - +
| 小明 | 今天  | 干了啥 |
+ - - + - - - + - - - +
| 就  | 吃饭  |       |
+ - - + - - - + - - - +
| 是  | 打豆豆 |       |
+ - - + - - - + - - - +
| 他  |       |       |
+ - - + - - - + - - - +
```

似乎哪里偏了……

**Comment by Artoria2e5** [comment]
> 平台不一样，中文宽度还不同

那属于 platform bug。认真的。

**Comment by leviathan0992** [comment]
呃。。。

**Comment by Arthur2e5** [comment]
……手动 `_pad_string` 都对的……传入 length * 2 难道会变…………？

**Comment by Arthur2e5** [comment]
……length 还真一直在变……

**Comment by Arthur2e5** [comment]
然而每列里面的 length 该一样啊……

**Comment by leviathan0992** [comment]
当初觉得代码简单，没有写注释，哔gou，现在都不知道自己写的是啥。。

**Comment by Artoria2e5** [comment]
I know where the problem is.. AttributesLength......

**Comment by leviathan0992** [comment]
这个是为了知道每列最宽的字符宽度是多少啊

**Comment by leviathan0992** [comment]
我把每列最宽的宽度纪录下来了

**Comment by Artoria2e5** [comment]
那边我没有 __disp_width。

**Comment by Artoria2e5** [comment]
现在修了，等一分钟推。。

```
Arthur2e5@wangmingye ~/GH/Pylsy
$ python2
Python 2.7.10 (default, Jun  1 2015, 18:05:38)
[GCC 4.9.2] on cygwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from pylsy import *
>>> t = pylsytable([u'name', u'love'])
>>> t.add_data('name', [u'小明', 'Alice'])
>>> t.add_data('loce', [u'搞比利！', u'啊'])
>>> print t.__str__()
+ - - - - - + - - - - +
|   name    |  love   |
+ - - - - - + - - - - +
|   小明    |         |
+ - - - - - + - - - - +
|   Alice   |         |
+ - - - - - + - - - - +

>>> t.add_data('love', [u'搞比利！', u'啊'])
>>> print t.__str__()
+ - - - - - + - - - - - - - - +
|   name    |      love       |
+ - - - - - + - - - - - - - - +
|   小明    |    搞比利！     |
+ - - - - - + - - - - - - - - +
|   Alice   |       啊        |
+ - - - - - + - - - - - - - - +

>>> t.AttributesLength
[5, 8]


```

image

**Comment by leviathan0992** [comment]
good!

**Comment by Artoria2e5** [comment]
Achievement gained: first python pr merged.

File: pylsy/pylsy.py [hunk 3]
```diff
@@ -81,31 +82,38 @@ def _print_head(self):
...
-                else:
-                    start = 0
-                    end = value_length - start + 1
-                    for space in range(end):
-                        self.StrTable += " "
+                self.StrTable += self._pad_string(val, length * 2)
             self.StrTable += "|"+'\n'
             self._print_divide()
 
+    def _disp_width(self, pwcs, n=None):
```
Comment (Artoria2e5) [inline pylsy/pylsy.py:92]: If I want to make this code short (and perhaps more Pythonic.. hey what is Pythonic?):

``` Python
    def _disp_width(self, pwcs, n=None):
        end = len(pwcs) if n is None else n
        idx = slice(0, end)
        return sum([max(0, wcwidth(char)) for char in pwcs[idx]])
```

Not wrapping `idx` inside since it's a func call (internal though)..

**Comment by MrLYC** [comment]
awesome!

File: pylsy/pylsy.py [hunk 4]
```diff
@@ -1,17 +1,24 @@
...
 class pylsytable(object):
 
     def __init__(self, attributes):
         self.StrTable = ""
-        self.Attributes = attributes
         self.Table = []
         self.AttributesLength = []
-        self.Cols_num = len(self.Attributes)
         self.Lines_num = 0
+        if type(attributes) != list:
```
Comment (bcho) [inline pylsy/pylsy.py:18]: It's better to use `isinstance(attributes, (tuple, list))`.

**Comment by bcho** [comment]
Actually there is no need to use wcwidth, `unicodedata` can do the same job here. [FYI](https://github.com/Robpol86/terminaltables/blob/master/terminaltables/width_and_alignment.py#L28-L35)

BTW, tests can ensure do things right for corner cases , please add some.

**Comment by Artoria2e5** [comment]
@bcho I hope not to limit things in the scope of some `width = 2 iff char is East Asian` assumption. Corner cases for non-printing chars......I don't even think these things should be passed there. I only changed it since `-1` will make things act way too ridiculously.

I am quite lazy these days so you can make a PR you know…
